src/interface/state_descriptor.cpp

```diff
@@ -320,47 +320,59 @@ std::ostream &operator<<(std::ostream &os, const StateDescriptor &sd) {
 // Takes all packages and combines them into a single state descriptor
 // containing all variables with conflicts resolved.  Note the new
 // state descriptor DOES not have any of its function pointers set.
-std::shared_ptr<StateDescriptor> ResolvePackages(Packages_t &packages) {
+std::shared_ptr<StateDescriptor>
+StateDescriptor::CreateResolvedStateDescriptor(Packages_t &packages) {
   auto state = std::make_shared<StateDescriptor>("parthenon::resolved_state");
 
   // The workhorse data structure. Uses sets to cache which variables
   // are of what type.
-  DependencyTracker var_tracker;
+  DependencyTracker dense_tracker;
+  DependencyTracker sparse_tracker;
   DependencyTracker swarm_tracker;
   // closures that provide functions for DependencyTracker
-  CellVarProvider cvar_provider(state);
-  SparseProvider sparse_provider(packages, state);
+  DenseFieldProvider dense_field_provider(state);
+  SparsePoolProvider sparse_pool_provider(packages, state);
   SwarmProvider swarm_provider(packages, state);
 
   // Add private/provides variables. Check for conflicts among those.
   // Track dependent and overridable variables.
   for (auto &pair : packages.AllPackages()) {
-    auto &name = pair.first;
+    const auto &name = pair.first;
     auto &package = pair.second;
-    // sort
-    var_tracker.CategorizeCollection(name, package->AllFields(), &cvar_provider);
-    for (auto &p2 : package->AllSparseFields()) { // sparse
-      auto &var = p2.first;
-      auto &mdict = p2.second;
-      for (auto &p3 : mdict) {
-        auto &metadata = p3.second;
-        var_tracker.Categorize(name, var, metadata, &sparse_provider);
+
+    // make metadata dictionary of dense variables and sparse pools (using the shared
+    // metadata, which contains the role information)
+    Dictionary<Metadata> dense_dict, sparse_dict;
+
+    for (const auto itr : package->AllFields()) {
+      if (!itr.second.IsSet(Metadata::Sparse)) {
+        dense_dict.insert({itr.first.label(), itr.second});
       }
     }
+
+    for (const auto itr : package->AllSparsePools()) {
+      sparse_dict.insert({itr.first, itr.second.shared_metadata()});
+    }
+
+    // sort
+    dense_tracker.CategorizeCollection(name, dense_dict, &dense_field_provider);
+    sparse_tracker.CategorizeCollection(name, sparse_dict, &sparse_pool_provider);
```

Comment:
Great question! I improved the detection mechanism a bit for sparse/dense variables of the same name. This all happens at the individual `StateDescriptor` level. Here's how it currently works:
- If you add a dense field "foo", subsequently trying to add a sparse pool with base name "foo" will return false (not throw, just like when you try to add another dense field "foo")
- Works the same in reverse: if you add a sparse pool "foo", trying to add another sparse pool or dense field "foo" will both return false and not add it
- If you add a sparse pool "foo" with sparse IDs 1, 2, then trying to add dense fields "foo_1" or "foo_2" will return false. However, adding "foo_3" will succeed
- Similarly, if you add a dense field "foo_15", you can add a sparse pool "foo" if it doesn't include the index 15, if it does, it will throw, because it would leave the `StateDescriptor` in an invalid state with some sparse fields "foo_x" added but "foo_15" being a dense field and the sparse "foo_15" not added

See the last few lines in `tst/unit/test_state_descriptor.cpp`.

I'm checking the dependency resolution between packaging with one having dense "foo_x" and the other sparse "foo_x".

Reply:
Cool! This all seems reasonable.

Reply:
So, this turned out to be a bigger issue. I didn't realize that the different dependency trackers operate completely independently. I.e. if we have a provided dense field "foo" and a provided sparse pool or swarm called "foo", the dependency resolution machinery happily tries to add both to the resolved state. This is ok for fields and swarms, since they are completely different entities, but it's not ok for dense fields and sparse pools, because they both result in `CellVariable`. The dependency resolution didn't catch this because it didn't check for it and when it added the second field/pool it didn't check that it was actually added (the `Add` function just returns false, signaling that it wasn't added, but the dependency machinery didn't check the return value).

I've now combined the dense field and sparse pool dependency tracker into one tracker and now we can detect if someone tries to provide a sparse pool and dense field with the same names. However, I *think* a consequence of this is that dense fields and sparse pools could now satisfy requirements for each other and override each other. Should we detect and not allow this?

Reply:
Good catch! 

> However, I think a consequence of this is that dense fields and sparse pools could now satisfy requirements for each other and override each other. Should we detect and not allow this?

Hmmm interesting... I don't know the answer to this... I could go either way. The path of least resistance is probably to leave it? What do others think? @jdolence @pgrete @forrestglines ?

Reply:
I'm inclined to leave it like this (at least for now). If it becomes a problem down the line, we'll address it

Reply:
Works for me